### ISSUE TITLE
create an sstableloader alternative

### ISSUE THREAD

**Issue by tchaikov** [scylladb/scylladb#14857]
this issue is part of #14856. the goal is to fill the gap between the load-n-stream and sstableloader.

despite that load-n-stream is superior than sstableloader in the sense that it is much more efficient, we cannot use load-n-stream on a random node. load-n-stream can only run on a node which is a member of the cluster. while sstableloader can run on any machine as long as it has connection to the cluster. and in theory, it can run on anywhere (quoting the slogan 20 years ago).

to be feature parity, we need to find a way allowing the cluster to consume the sstables from "anywhere".  one way is to rework `sstables_loader::load_new_sstables()` so that it is able to perform following operations remotely

1. list the sstables 
2. read mutation fragments as a stream from a set of the sstables (think about a generator which yields `mutation_fragment` with pagination support)

**Comment by avikivity** [comment]
We can have load'n'stream access a remote sstable via S3. It's not completely general but it covers many use cases.

Procedure:
1. upload sstables to S3
2. ask each node to load'n'stream some of them
3. eat cookies

**Comment by mykaul** [comment]
> 1. upload sstables to S3

Who sets the metadata for it? I assume it's not just using AWS CLI to upload an sstable to a bucket?

**Comment by tchaikov** [comment]
> > 1. upload sstables to S3
> 
> Who sets the metadata for it? I assume it's not just using AWS CLI to upload an sstable to a bucket?

what do you mean by "metadata"? i think both sstableloader and load'n'stream require the pre-populated keyspace and table.

**Comment by mykaul** [comment]
> > > 1. upload sstables to S3
> > 
> > 
> > Who sets the metadata for it? I assume it's not just using AWS CLI to upload an sstable to a bucket?
> 
> what do you mean by "metadata"? i think both sstableloader and load'n'stream require the pre-populated keyspace and table.

Ownership table population, for example.

**Comment by tchaikov** [comment]
i see. so in other words, how can scylla tell which sstables it should load?

i think this problem is the exactly the one we should attack when designing the use case of node replacement with the object/cached store backend. Benny proposed a tagging based machinery allowing load'n'stream (or other functionalities) to filter the listed sstables by its tag, in pseudocode:

```py
class sstables_loader:
  def load_and_stream(db, ks_name, cf_name, node_id):
    # node_id: is the previous owner of the sstables to be inherited by the successor 
    table = db.get_table(ks_name, cf_name)
    storage_options = table.storage_options()
    storage = sstable.storage(storage_options)
    sstables = storage.get_sstables_from_upload_dir(ks_name, cf_name, node_id)
    self.container.invoke_on_all(lambda loader: load_and_stream(sstables))

class storage:
  def get_sstables_from_upload_dir(ks_name, cf_name, node_id):
    raise NotImplementedError()

class file_storage(storage):
  def get_sstables_from_upload_dir(ks_name, cf_name, node_id):
    # enumerate and load sstables on all shards in parallel
    # see distributed_loader::get_sstables_from_upload_dir()

class s3_storage(storage):
  def get_sstables_from_upload_dir(ks_name, cf_name, node_id):
    # we assume that all the data sstable should be attached with the "owner"
    # tag once it is sealed or fully uploaded
    sstables = self.bucket.list(prefix='upload').filter_by_tag(lambda tags: tags['owner'] == node_id)
    # assign the found sstables to their shards
```

Avi's suggestion is great. but IMO, it is missing some details. i am trying to write down a possible implementation here to progress the discussion.

**Comment by denesb** [comment]
I think with the integrated backup-restore via S3 (https://github.com/scylladb/scylladb/pull/20305) we now have the Scylla-side ready.
We need https://github.com/scylladb/scylladb/issues/20538 and some glue around it to produce the expected layout in S3 and then we have our `sstableloader` alternative.

**Comment by denesb** [comment]
After some further reading, I think this procedure could work:
* Create table on new cluster.
* Obtain table name and table uuid: `cqlsh> select table_name, id from system_schema.tables where keyspace_name = '$keyspace_name' and table_name = '$table_name'`.
* Upload sstables to appropriate S3 backup location: `scylla sstable upload --endpoint=$endpoint --bucket=$bucket --prefix=${table_name}-${table_id}/$snapshot_name /path/to-sstables`. Use `--concurrency` or spin up multiple processes (needs division of work) to increase concurrency and reduce wait-time.
* Start restore from S3: `nodetool restore --keyspace=$keyspace_name --table=$table_name --endpoint=$endpoint --bucket=$bucket --snapshot=$snapshot_name`

Where:
* `$keyspace_name`, `$table_name` and `$table_uuid` are properties of the table.
* `$endpoint` - a valid endpoint, which is present in the S3 configuration.
* `$bucket` and `$snapshot_name` - can be any arbitrary value, it just has to be consistently used throughout the procedure. Maybe there are somerestrictions on the naming patterns, that I don't know yet. There will probably be some best-practices to ensure `$bucket` doesn't collide with ones created by ScyllaDB cluster (or simply clean up bucket after restore).

**Comment by denesb** [comment]
Just noting here, that we are also considering just removing sstableloader without any direct successors: https://github.com/scylladb/scylladb/pull/20740#issuecomment-2405831267.

**Comment by bhalevy** [comment]
how does that fit the integrated restore design?
IIUC today scylla-manager uses rclone or a similar tool to copy files from object storage to a local uploads/ directory and then it invokes `nodetool refresh --load-and-stream`.

As @avikivity said in https://github.com/scylladb/scylladb/issues/14857#issuecomment-1653450033, we can extend load-and-stream to load the data directly from S3 without having to copy it first.  And the sstable doesn't even need to be registered in the ownership table since we just stream its contents and never add it to the table sstables set.

Cc @regevran @xemul @Michal-Leszczynski @karol-kokoszka

**Comment by regevran** [comment]
Now we don't need two steps for restore:
* The Scylla API combines download and load&stream into a single API call.
* [load and stream from S3](https://github.com/scylladb/scylladb/blob/master/sstables_loader.cc#L492-L504).

I am not sure we need alternatives.
@denesb, @tchaikov  - we should discuss this.

**Comment by bhalevy** [comment]
To make it clear, I think it's fine to have an alternative to sstableloader and to support the mative object storage stack in scylla sstable tool (especially for dump).
Just that I don't think we need it for restore.

**Comment by denesb** [comment]
We have decided to remove sstableloader in 6.3.

We should add a notice to 6.2 release notes about the upcoming deprecation.

Closing as wontfix.

@tzach @mykaul
